Use go-cmp instead of reflect.DeepEqual

hslatman opened the pull request:
As discussed in https://github.com/cweill/gotests/pull/99, the branch had to be cleared from merge conflicts before `go-cmp` support could be integrated. That's what I did in this PR.

hslatman commented:
@googlebot I consent.

butuzov commented:
Hi @hslatman can you merge develop into your feature? I believe I brought some conflicts to your code.

hslatman commented:
@butuzov I've merged develop and regenerated the bindata.

While the tests run successfully on GitHub Actions I have two failing tests locally. Is there something special I need to do to to run the tests locally successfully too?

butuzov commented:
I would suggest do not run them with `-race` or concurrently [atm](https://github.com/cweill/gotests/issues/154).

@cweill can you also check this PR?

butuzov commented on gotests_test.go:
```diff
@@ -918,7 +919,7 @@ func TestGenerateTests(t *testing.T) {
 			}
 			if got := string(gts[0].Output); got != tt.want {
-				t.Errorf("%q. GenerateTests(%v) = \n%v, want \n%v", tt.name, tt.args.srcPath, got, tt.want)
+				t.Errorf("%q. GenerateTests(%v) = diff=%v", tt.name, tt.args.srcPath, cmp.Diff(got, tt.want))
```
Would it be possible to create maybe a custom reporter so we can get `got/want` with a lines of the diff instead `-/+` ?

also `%v` now can be `%s`

hslatman replied:
The custom reporter approach certainly looks interesting. So far I just updated @mekegi's code, but I could give integrating the custom reporter a try.

So your goal would be to print the output of the custom reporter instead of the default cmp.Diff(x, y), right? It seems doable for this specific line of code, i.e. in the `gotests_test.go` it should be fairly trivial to add. 

Do you mean to also use the custom reporter for the generated tests? And by extension, making that configurable by the user of gotests too? Or am I way off and did you mean to just put it in `gotests_test.go`?

hslatman replied:
Taking the example DiffReporter and changing the lines to this:

```go
var r reporter.DiffReporter
if got := string(gts[0].Output); !cmp.Equal(got, tt.want, cmp.Reporter(&r)) {
    t.Errorf("%q. GenerateTests(%v) = diff=%s", tt.name, tt.args.srcPath, r.String())
    outputResult(t, tmp, tt.name, gts[0].Output)
}

// basic Report() function, slightly altered from the example
func (r *DiffReporter) Report(rs cmp.Result) {
	if !rs.Equal() {
		vx, vy := r.path.Last().Values()
		r.diffs = append(r.diffs, fmt.Sprintf("\n\nGot:\n\n%+v\n\nWant:\n\n%+v\n", vx, vy))
	}
}
```

Results in the following output:

```bash
    gotests_test.go:924: "Receiver is indirect imported struct". GenerateTests(testdata/test037.go) = diff=

        Got:

        package testdata

        import "testing"

        func Test_someIndirectImportedStruct_Foo037(t *testing.T) {
        	tests := []struct {
        		name string
        		smtg *someIndirectImportedStruct
        	}{
        		// TODO: Add test cases.
        	}
        	for _, tt := range tests {
        		tt.smtg.Foo037()
        	}
        }


        Want:

        package testdata

        import "testing"

        func Test_someIndirectImportedStruct_Foo037(t *testing.T) {
        	tests := []struct {
        		name string
        		smtg *someIndirectImportedStruct
        	}{
        		// TODO: Add test cases.
        	}
        	for _, tt := range tests {
        		smtg := &someIndirectImportedStruct{}
        		smtg.Foo037()
        	}
        }

    gotests_test.go:958: /var/folders/m1/h8xf_67j5fqccwr1277f3_zc0000gn/T/gotests_test230212921/receiver_is_indirect_imported_struct.go
```

This doesn't seem really useful, but it may be possible to customize it down to the actual lines that are different. 

If I understand you correctly, you want to show just got/want and the corresponding lines, correct? I find the -/+ quite easy to use, so I'm not sure if we should drop that, unless you want it to become configurable for a user of gotests?

butuzov replied:
I was talking only about generated tests and you are right without granular line detection it does not seem really useful.

cweill replied:
+1 to keeping the -/+. Without them it's very difficult to know what the delta is.

butuzov reviewed (commented):
Maybe we `%s` placeholders can be used instead of `%v` in all places now?

butuzov commented on gotests.go:
```diff
@@ -16,6 +16,13 @@ import (
 	"github.com/cweill/gotests/internal/output"
 )
 
+var (
+	cmpImport = &models.Import{
+		Name: "",
+		Path: `"github.com/google/go-cmp/cmp"`,
+	}
+)
```
minor: is there any other data it should be grouped in the future?

```suggestion
var cmpImport = &models.Import{
	Name: "",
	Path: `"github.com/google/go-cmp/cmp"`,
}
```

butuzov reviewed (commented):
really minor nit

butuzov commented:
ping @cweill

cweill commented on .github/workflows/go.yml:
```diff
@@ -35,6 +35,9 @@ jobs:
       run: go get -v ./...
 
+    - name: Get go-cmp
+      run: go get github.com/google/go-cmp/cmp
```
Why do we need to manually fetch this package? If a user forgets to do this will gotests not work for them? IMO `go get -v ./...` should cover installing this dependency.

hslatman replied:
Without this explicit call the tests will fail in GitHub Actions for Go 1.9 - 1.12:

```
Run go test -v ./...
  go test -v ./...
  shell: /usr/bin/bash -e {0}
  env:
    GOPATH: /home/runner/work/gotests/gotests
    GO111MODULE: auto
    GOROOT: /opt/hostedtoolcache/go/1.12.17/x64
# github.com/cweill/gotests
FAIL	github.com/cweill/gotests [setup failed]
gotests_test.go:16:2: cannot find package "github.com/google/go-cmp/cmp" in any of:
	/opt/hostedtoolcache/go/1.12.17/x64/src/github.com/google/go-cmp/cmp (from $GOROOT)
	/home/runner/work/gotests/gotests/src/github.com/google/go-cmp/cmp (from $GOPATH)
FAIL	github.com/cweill/gotests/gotests [setup failed]
```
 
Other solutions to make this work may exist. The `GO111MODULE` env var could be set to `on`, but that'll likely only work for Go 1.11 and up. To me this seemed the easiest solution to make it work across versions without having to make the CI configuration more complex.

I tried the `go-cmp` version of `gotests` locally on a sample project. It resulted in a `_test.go` file with `github.com/google/go-cmp/cmp` in the import section. Running the tests for the first time results in `go-cmp` being inserted into go.mod. The go.mod of gotests itself does contain `go-cmp`, so that should work correctly too.

cweill replied:
My concern is for who don't have go-cmp installed, will this change break them unless they manually install the package? Is there any way we can install it automatically when they install gotests?

cweill commented:
@mekegi Could you please reply `@googlebot I consent.` to this PR?

mekegi commented:
@googlebot I consent

hslatman commented:
I'm curious to know if the output format is OK for you as it is now. It looks like this now:

```
        diff=  (*client.Client)(
        - 	&{
        - 		key:                              "key",
        - 		httpClient:                    &http.Client{Timeout: s"30s"},
        - 		baseURL:                     s"https://example.com/api",
        - 		contentTypeHeader:   "application/json",
        - 		acceptHeader:            "application/json",
        - 	},
        + 	nil,
          )
```

One other thing that I found about is that having unexported fields in structs will result in panics when running tests if they are not explicitly allowed, as mentioned in the third point in the `go-cmp` repo:

```
_Unlike reflect.DeepEqual, unexported fields are not compared by default; they result in panics unless suppressed by using an Ignore option (see cmpopts.IgnoreUnexported) or explicitly compared using the AllowUnexported option._
```

One would thus have to change the generated `!cmp.Equal(got, tt.want)` and cmp.Diff(got, tt.want) to something like the below (or one of the other options mentioned above):

```go
if !cmp.Equal(got, tt.want, cmpopts.IgnoreUnexported(StructUnderTest{})) {
	t.Errorf("%q. New() = %v, want %v\ndiff=%s", tt.name, got, tt.want, cmp.Diff(got, tt.want, cmpopts.IgnoreUnexported(StructUnderTest{})))
}
```

cweill commented:
> I'm curious to know if the output format is OK for you as it is now. It looks like this now:

Looks good to me.

> One other thing that I found out about is that having unexported fields in structs will result in panics when running tests if they are not explicitly allowed, as mentioned in the third point in the go-cmp repo:
> 
> Unlike reflect.DeepEqual, unexported fields are not compared by default; they result in panics unless suppressed by using an Ignore option (see cmpopts.IgnoreUnexported) or explicitly compared using the AllowUnexported option.
> 
> One would thus have to change the generated !cmp.Equal(got, tt.want) and cmp.Diff(got, tt.want) to something like the below (or one of the other options mentioned above):

I think we should stick to what's simplest, if that means it can panic, at least users can modify the generated boilerplate code to match what they need.

A bigger thing is, I'm not sure if using go-cmp by the default is the right approach. The Golang philosophy is to use the stdlib for most things, and only third-party packages if necessary. Since go-cmp is a third-party package, I'd suggest flag protecting it behind a flag like `--use_gocmp`, and using DeepEqual by default. This way gotests users can opt into go-cmp only if they want it. In a future release we can consider making go-cmp the default.

mrsufgi commented:
Any updates on this?

hslatman commented:
It's been on the backburner for me for quite a while, but I may be able to give this some attention again sometime soon.

hslatman commented:
@cweill: can you start the tests?

I've added the `-use_go_cmp` flag, so usage is now conditional. I've reverted old test files to using `reflect.DeepEqual` again and made copies of relevant tests to use `cmp.Equal` and `cmp.Diff`.

hslatman commented:
@cweill: I've added another commit that removes some old Go versions from the CI, which makes the tests pass again: https://github.com/hslatman/gotests/actions/runs/1976091466. 

The error I got is the same as in the issue that @davidhsingyuchen reported in https://github.com/cweill/gotests/issues/168.

edgio-mwodrich commented:
Digging into cmp.Diff a little, it runs cmp.Equal internally itself, and will produce string output if and only if cmp.Equal is false (if they ever disagree, it panics), so instead of 

```go
if !cmp.Equal(got, tt.want, opts) {
    t.Errorf("Function diff(got, want):\n%s", cmp.Diff(got, tt.want, opts))
}
```

a simplification for performance and maintainability I've adopted in gotests I've converted to use cmp.Diff is to do something like

```go
if diff := cmp.Diff(got, tt.want, opts); diff != "" {
    t.Errorf("Function diff(got, want):\n%s", diff)
}
```

to avoid redundant runs of cmp.Equal and having to keep separate cmp.Equal and cmp.Diff invocations in sync, especially when using cmpopts.

Either way, I'd be very excited to use this type of functionality in my use of gotests, and I hope that folks might have time to move it forward.

cweill commented:
Thank you for this contribution and for resolving the merge conflicts from PR #99!

This feature has been implemented and released in **v1.7.1**! 🎉

The implementation uses a `-use_go_cmp` flag that allows users to opt-in to using `cmp.Equal()` instead of `reflect.DeepEqual()`, with `cmp.Diff()` in error messages for better test failure output.

**Usage:**
```bash
gotests -use_go_cmp -all -w example.go
```

**Generated test code:**
```go
if !cmp.Equal(tt.want, got) {
    t.Errorf("Foo() = %v, want %v\ndiff=%s", got, tt.want, cmp.Diff(tt.want, got))
}
```

The feature works exactly as you intended, with automatic import of `github.com/google/go-cmp/cmp`.

Thanks for your work on this - your PR was very helpful as a reference for the implementation!

See the release: https://github.com/cweill/gotests/releases/tag/v1.7.1

hslatman commented:
Awesome! 🎉
